src/analytics/src/main/scala/osmesa/analytics/stats/WayLengthCalculator.scala

```diff
@@ -37,9 +45,38 @@ object WayLengthCalculator {
         }.
         toMap
 
-    // To be safe and not blow up the job on bad data, if there are any
-    // nodes the way is pointing to that we don't have in `nodes`, just skip
-    // this version of the way.
+    val (wayInstants, waysArr) =
+      ways.
+        toArray.
+        sortBy(_._3).
+        map { case (nodeIds, changeset, instant, topicMap) =>
+          (instant, (nodeIds, changeset, topicMap))
+        }.
+        unzip
+
+    /** Gets the nodes and topics for a way given an instant. The way chosen will be
+      * the most recent way that is <= the instant.
+      * If the instant is before all instants of the way, return None.
+      */
+    def getWayForInstant(instant: Long): Option[(Array[Long], Long, Map[StatTopic, Boolean])] = {
+      val idx = {
+        // index of the search key, if it is contained in the array; otherwise, (-(insertion point) - 1).
+        val i = Arrays.binarySearch(wayInstants, instant)
+        if(i >= 0) { i }
+        else {
+          ~i - 1
+        }
+      }
+
+      if(idx < 0) { None }
+      else { Some(waysArr(idx)) }
+    }
+
+    /** Gets the length of this way at an instant.
+      *  To be safe and not blow up the job on bad data, if there are any
+      * nodes the way is pointing to that we don't have in `nodes`, just skip
+      * this version of the way.
```

Comment:
Should we dump these somehow? Maybe even just as a count (per-changeset, to help track them down).

Reply:
Would be good to, we should have some sort of logging that writes out bad geoms. That can be a TODO